File: base/src/main/java/edu/hitsz/aircraft/EliteEnemy.java
package edu.hitsz.aircraft;

import edu.hitsz.animate.AnimateContainer;
import edu.hitsz.application.Game;
import edu.hitsz.application.Main;
import edu.hitsz.bullet.BaseBullet;
import edu.hitsz.Utils;
import edu.hitsz.prop.AbstractProp;
import edu.hitsz.prop.BloodPropFactory;
import edu.hitsz.prop.BombPropFactory;
import edu.hitsz.prop.BulletPropFactory;
import edu.hitsz.vector.Vec2;

import java.util.LinkedList;
import java.util.Random;

/**
 * 精英敌机，可以射击
 *
 * @author hitsz
 */
public class EliteEnemy extends AbstractAircraft {

    final Random random = new Random();

    public EliteEnemy(Vec2 posInit, AnimateContainer animateContainer, int hp) {
        super(posInit, animateContainer, hp, 100);
    }

    @Override
    public void forward() {
        super.forward();
        // 判定 y 轴向下飞行出界
        if (getLocationY() >= Main.WINDOW_HEIGHT) {
            vanish();
        }
    }

    @Override
    public LinkedList<BaseBullet> shoot() {
        return Utils.letEnemyShoot(getPosition());
    }

    @Override
    public LinkedList<AbstractProp> dropProps() {
        double select = random.nextDouble();
        double probability = 0.3;
        LinkedList <AbstractProp> props = new LinkedList<>();
        if (Utils.isInRange(select, 0, probability)) {
            props.add(new BloodPropFactory(getPosition()).create());
        } else if (Utils.isInRange(select, probability, probability * 2)) {
            props.add(new BombPropFactory(getPosition()).create());
        } else if (Utils.isInRange(select, probability * 2, probability * 3)) {
            props.add(new BulletPropFactory(getPosition()).create());
        }
        // else no props.
        return props;
    }
}

File: src/main/scala/work/chiro/game/aircraft/EliteEnemy.scala
package work.chiro.game.aircraft

import work.chiro.game.GlobalConfigLoader.config
import work.chiro.game.animate.{AnimateContainer, AnimateLinear, AnimateNonLinear, AnimateVectorType}
import work.chiro.game.application.ImageResourceFactory
import work.chiro.game.basic.PositionType.{Position, Scale, Size, SizeDouble}
import work.chiro.game.basic.{AbstractObjectFactory, Vec2Double}
import work.chiro.game.bullet.EnemyBullet
import work.chiro.game.utils.{getNewFlightPosition, getTimeMills}

/**
 * 普通敌机，不可攻击
 *
 * @author chiro2001
 * @param posInit          敌机初始位置
 * @param animateContainer 动画容器
 * @param hpInit           初始血量
 */
class EliteEnemy(posInit: Position,
                 animateContainer: AnimateContainer,
                 hpInit: Int,
                 sizeInit: Option[SizeDouble] = None,
                 rotationInit: Option[Scale] = None)
  extends AbstractAircraft(posInit, animateContainer, hpInit, sizeInit = sizeInit, rotationInit = rotationInit) {

  val power = 10

  override def forward() = {
    super.forward()
    // 判定 y 轴向下飞行出界
    if (getLocationY >= config.window.playHeight) vanish()
  }

  override def shoot() = List({
    val posNew = new Position(getLocationX, getLocationY)
    new EnemyBullet(posNew, new AnimateContainer(List(
      new AnimateNonLinear(posNew, HeroAircraft.getPositionInstance, AnimateVectorType.PositionLike.id, getTimeMills, 3000, willStop = false)
    )), power)
  })

  override def getImage = EliteEnemy.getImage
}

object EliteEnemy extends ImageResourceFactory with AbstractObjectFactory {
  override def getImageCachedPath = "images/elite.png"

  def create() = {
    val positionEnemyNew = getNewFlightPosition(EliteEnemy.getImage.getWidth)
    val newRotation = new Scale
    new EliteEnemy(
      positionEnemyNew, new AnimateContainer(List(
        new AnimateLinear(positionEnemyNew, new Position(0, 0.1),
          AnimateVectorType.PositionLike.id, getTimeMills, 0)
      )), 30, rotationInit = Some(newRotation)
    )
  }
}
